src/sync/barrier.rs

```diff
@@ -22,9 +22,19 @@ impl BarrierWaitResult {
 
```

Comment:
Could you write a comment like the one we have for Condvar explaining how/why the implementation works?

https://github.com/awslabs/shuttle/blob/94b9d8bc87b0cb11c213cccdb45240e32a16e6eb/src/sync/condvar.rs#L40

It should probably be a doc comment on `BarrierState`? (the Condvar one should probably have been a doc comment on `CondvarState`).